src/experimental/collection_new_iterable_model/shared/iterator/RedBlackTreeIterator.hpp

```diff
@@ -0,0 +1,397 @@
+//
+// Created by stefan on 16.08.2022.
+//
+
+#ifndef __CDS_EX_SHARED_RED_BLACK_TREE_ITERATOR_HPP__
+#define __CDS_EX_SHARED_RED_BLACK_TREE_ITERATOR_HPP__
+
+namespace cds {                 // NOLINT(modernize-concat-nested-namespaces)
+    namespace experimental {
+
+
+        namespace __hidden {    // NOLINT(modernize-concat-nested-namespaces, bugprone-reserved-identifier)
+            namespace __impl {  // NOLINT(bugprone-reserved-identifier)
+                template <
+                        typename __TElementType,                                                                // NOLINT(bugprone-reserved-identifier)
+                        typename __KeyType,                                                                     // NOLINT(bugprone-reserved-identifier)
+                        cds :: utility :: ExtractorFunction < __TElementType, __KeyType > __keyExtractor,       // NOLINT(bugprone-reserved-identifier)
+                        cds :: utility :: ComparisonFunction < __TElementType > __keyLowerComparator,           // NOLINT(bugprone-reserved-identifier)
+                        cds :: utility :: ComparisonFunction < __TElementType > __keyEqualsComparator,          // NOLINT(bugprone-reserved-identifier)
+                        cds :: utility :: DestructorFunction < __TElementType > __nodeDestructor                // NOLINT(bugprone-reserved-identifier)
+                > class __RedBlackTree;                                                                         // NOLINT(bugprone-reserved-identifier)
+            }
+        }
+
+        template < typename __ElementType >     // NOLINT(bugprone-reserved-identifier)
+        class AbstractTreeIterator : public meta :: BidirectionalIterator {
+            template <
+                    typename __TElementType,                                                                // NOLINT(bugprone-reserved-identifier)
+                    typename __KeyType,                                                                     // NOLINT(bugprone-reserved-identifier)
+                    cds :: utility :: ExtractorFunction < __TElementType, __KeyType > __keyExtractor,       // NOLINT(bugprone-reserved-identifier)
+                    cds :: utility :: ComparisonFunction < __TElementType > __keyLowerComparator,           // NOLINT(bugprone-reserved-identifier)
+                    cds :: utility :: ComparisonFunction < __TElementType > __keyEqualsComparator,          // NOLINT(bugprone-reserved-identifier)
+                    cds :: utility :: DestructorFunction < __TElementType > __nodeDestructor                // NOLINT(bugprone-reserved-identifier)
+            > friend class cds :: experimental :: __hidden :: __impl :: __RedBlackTree;                     // NOLINT(bugprone-reserved-identifier)
```

Comment:
the __keyExtractor, __keyLowerComparator, __keyEqualsComparator, __nodeDestructor parameters can remain unnamed.